src/structures/ComponentInteractionCollector.js

```diff
@@ -0,0 +1,145 @@
+
+/**
+ * @typedef {CollectorOptions} ComponentInteractionCollectorOptions
+ * @property {number} max The maximum total amount of interactions to collect
+ * @property {number} maxComponents The maximum number of buttons to collect
+ * @property {number} maxUsers The maximum number of users to interact
+ */
+
+/**
+ * Collects interaction on message buttons.
+ * Will automatically stop if the message (`'messageDelete'`),
+ * channel (`'channelDelete'`), or guild (`'guildDelete'`) are deleted.
+ * @extends {Collector}
+ */
+class ComponentInteractionCollector extends Collector {
+  /**
+   * @param {Message} message The message upon which to collect button interactions
+   * @param {CollectorFilter} filter The filter to apply to this collector
+   * @param {ComponentInteractionCollectorOptions} [options={}] The options to apply to this collector
+   */
+  constructor(message, filter, options = {}) {
+    super(message.client, filter, options);
+
+    /**
+     * The message upon which to collect button interactions
+     * @type {Message}
+     */
+    this.message = message;
+
+    /**
+     * The users which have interacted to buttons on this message
+     * @type {Collection}
+     */
+    this.users = new Collection();
+
+    /**
+     * The total number of interactions collected
+     * @type {number}
+     */
+    this.total = 0;
+
+    this.empty = this.empty.bind(this);
+    this._handleChannelDeletion = this._handleChannelDeletion.bind(this);
+    this._handleGuildDeletion = this._handleGuildDeletion.bind(this);
+    this._handleMessageDeletion = this._handleMessageDeletion.bind(this);
+
+    this.client.incrementMaxListeners();
+    this.client.on(Events.INTERACTION_CREATE, this.handleCollect);
+    this.client.on(Events.MESSAGE_DELETE, this._handleMessageDeletion);
+    this.client.on(Events.CHANNEL_DELETE, this._handleChannelDeletion);
+    this.client.on(Events.GUILD_DELETE, this._handleGuildDeletion);
+
+    this.once('end', () => {
+      this.client.removeListener(Events.INTERACTION_CREATE, this.handleCollect);
+      this.client.removeListener(Events.MESSAGE_DELETE, this._handleMessageDeletion);
+      this.client.removeListener(Events.CHANNEL_DELETE, this._handleChannelDeletion);
+      this.client.removeListener(Events.GUILD_DELETE, this._handleGuildDeletion);
+      this.client.decrementMaxListeners();
+    });
+
+    this.on('collect', interaction => {
+      this.total++;
+      this.users.set(interaction.user.id, interaction.user);
+    });
+  }
+
+  /**
+   * Handles an incoming interaction for possible collection.
+   * @param {Interaction} interaction The interaction to possibly collect
+   * @returns {?Snowflake|string}
+   * @private
+   */
+  collect(interaction) {
+    /**
+     * Emitted whenever a reaction is collected.
+     * @event ComponentInteractionCollector#collect
+     * @param {ComponentInteraction} interaction The reaction that was collected
+     */
+    if (!interaction.isButton()) return null;
```

Comment:
```suggestion
    if (!interaction.isComponent()) return null;
```
I wasn't able to find `<Interaction>#isButton`, is this supposed to be `isComponent` instead?

Reply:
Yes, was only part of the way through refactoring this. Still a draft.